style: Add infrastructure to support lazy pseudo-elements

emilio opened the pull request:
This builds on top of #10815, so it's really just the last commit the one that should be reviewed.

I tried to apply the new infrastructure to servo, but failed (for now?).

The problem with it is that it'd require `ThreadSafeLayoutElement` to implement `selectors::Element`, which is a lot of work and might be racy (not totally sure about it though). Thus, I prefered to keep selectors eager until knowing that it's safe to do it.

r? @mbrubeck for style changes, @bholley for the geckolib changes (minimal for now, glue + a list of lazy PEs must be added)

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/10934)

<!-- Reviewable:end -->

emilio committed:
style: Support lazy pseudo-elements

These can't be supported in Servo as of right now, because I'm not
totally sure the accesses that should be done in layout would be
thread-safe.

It can be revisited later though.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/selector_matching.rs, components/style/servo.rs, components/style/traversal.rs, components/style/matching.rs, components/style/media_queries.rs, components/style/restyle_hints.rs, components/style/dom.rs, components/style/context.rs, components/style/selector_impl.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

emilio commented on components/layout/wrapper.rs:
```diff
@@ -774,13 +774,41 @@ pub trait ThreadSafeLayoutNode : Clone + Copy + Sized + PartialEq {
+                    PseudoElementCascadeType::Lazy => {
+                        panic!("Lazy pseudo-elements can't be used in Servo \
+                               since accessing the DOM tree during layout \
+                               could be unsafe.")
```
Not totally sure about this, @mbrubeck, do you know if it'd be safe for `ServoThreadSafeLayoutElement` to implement `selectors::Element`? (that is, getting parent, siblings...)

bholley replied:
I'm pretty sure it would not be, but I'll let @mbrubeck confirm. Might need to think a bit on how this fits into the Servo model.

mbrubeck replied:
Right, it is unsafe for a ThreadSafeLayoutNode to access its parents or siblings.

emilio replied:
We could still use a subset of selector-matching though (attributes + tag name at least), enough for avoiding doing the whole cascade for some private pseudos, like `::-servo-details-summary`, which only needs access to the `details` tag and the `open` attribute.

Would this be desirable (with appropriate docs of course), or is it just a lot of overhead? I think it can be worth it but I'd like to know your opinion.

mbrubeck replied:
That seems reasonable to me; I guess all it would require is an impl of `Element` that just returns `None` for things like `parent_element`.

emilio replied:
Yup, just did that and cleaned it up a bit, feel free to re-review the changes :)

bholley commented on ports/geckolib/selector_impl.rs:
```diff
@@ -89,6 +89,36 @@ pub enum PseudoElement {
+            MozTreeDropFeedback | MozSVGMarkerAnonChild |
+            MozSVGOuterSVGAnonChild | MozSVGForeignContent |
+            MozSVGText => true,
+            _ => false,
```
Seems like we should make this an ADT?

emilio replied:
Done

bholley commented:
geckolib changes look good modulo that one change.

emilio committed:
geckolib: Make anon-box PEs an ADT.

mbrubeck commented on components/layout/wrapper.rs:
```diff
@@ -1252,8 +1280,8 @@ impl<ConcreteNode> Iterator for ThreadSafeLayoutNodeChildrenIterator<ConcreteNod
-/// A wrapper around elements that ensures layout can only ever access safe properties and cannot
-/// race on elements.
+/// A wrapper around elements that ensures layout can only
+/// ever access safe properties and cannot race on elements.
```
Nit: Rust style is to wrap at 99 columns.  I don't care too deeply about exactly where comments wrap, but if you can configure your editor not to re-format things to 80 columns, it would make reviews a little easier.

mbrubeck commented on components/style/selector_impl.rs:
```diff
@@ -8,56 +8,83 @@ use selectors::Element;
     #[inline]
     fn each_eagerly_cascaded_pseudo_element<F>(mut fun: F)
         where F: FnMut(<Self as SelectorImpl>::PseudoElement) {
         Self::each_pseudo_element(|pseudo| {
```
Should be renamed to `each_eager_pseudo_element`.

mbrubeck commented on components/style/selector_impl.rs:
```diff
@@ -191,8 +226,8 @@ impl SelectorImplExt for ServoSelectorImpl {
     #[inline]
-    fn is_eagerly_cascaded_pseudo_element(pseudo: &PseudoElement) -> bool {
-        pseudo.is_eagerly_cascaded()
+    fn pseudo_element_cascade_type(pseudo: &PseudoElement) -> PseudoElementCascadeType {
```
You could get rid of this function, since it's just an alias for `pseudo.cascade_type()`.

mbrubeck commented on components/layout/wrapper.rs:
```diff
@@ -774,13 +774,41 @@ pub trait ThreadSafeLayoutNode : Clone + Copy + Sized + PartialEq {
+                        //     data.style_data.per_pseudo
+                        //         .insert(style_pseudo.clone(), new_style.unwrap())
+                        // }
+                    }
```
Please remove the commented-out code before merging.

mbrubeck commented:
The Servo changes are good; r=mbrubeck with the commented-out code removed.  Note: I left a couple of other review comments but then deleted them because they were incorrect.  (This is what I get for using GitHub comments instead of Reviewable.)  Please disregard any comments that no longer appear in the GitHub UI.

notriddle commented:
Now that #10815 is merged, should this commit be rebased on top of it (so that GitHub only lists one commit for this PR)?o

bholley commented:
Yes. I'm guessing @emilio will do that sometime when he's awake. :-)

emilio committed:
style: layout: Allow a lazy pseudo-element implementation in Servo.

emilio committed:
style: Minor documentation tweaks about lazy pseudo-elements

highfive commented:
New code was committed to pull request.

highfive commented:
New code was committed to pull request.

highfive commented:
New code was committed to pull request.

mbrubeck commented:
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 6 of 6 files at r1, 2 of 11 files at r2, 3 of 7 files at r19, 3 of 6 files at r23, 3 of 3 files at r24.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[components/layout/wrapper.rs, line 1360 \[r24\]](https://reviewable.io:443/reviews/servo/servo/10934#-KGrotUZTmHMlRn-bN5x:-KGrotU_pJYzVm39SPAj:2086453147) ([raw file](https://github.com/servo/servo/blob/1138b89920ec9b8414551c0e311689258b4d601f/components/layout/wrapper.rs#L1360)):*
Is it a programming error if this is ever called? If so, maybe change these from `warn!` to `panic!` or `debug_assert!(false, ...)`.

---

*[components/style/selector_impl.rs, line 117 \[r24\]](https://reviewable.io:443/reviews/servo/servo/10934#-KGrqGoZkEbEjmG6RvOz:-KGrqGo_ARls5H4f_2sk:1278718828) ([raw file](https://github.com/servo/servo/blob/1138b89920ec9b8414551c0e311689258b4d601f/components/style/selector_impl.rs#L117)):*
This `TODO` comment is done. \o/  (A note about `::selection` would still be appropriate.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10934#-:-KGrqkk-LYx7nBvTwPib:1458266033)*
<!-- Sent from Reviewable.io -->

highfive commented:
New code was committed to pull request.

emilio commented:
Review status: 8 of 9 files reviewed at latest revision, 6 unresolved discussions.

---

*[components/layout/wrapper.rs, line 811 \[r19\]](https://reviewable.io:443/reviews/servo/servo/10934#-KGYqChKT1aeyfKMTiXZ-r19-811:-KGrsUfkoZT9m-jLBWWa:-499146009) ([raw file](https://github.com/servo/servo/blob/0a8c6bc05844ba7cb132460743cf3d4415c03e5b/components/layout/wrapper.rs#L811)):*
Done.

---

*[components/layout/wrapper.rs, line 1360 \[r24\]](https://reviewable.io:443/reviews/servo/servo/10934#-KGrotUZTmHMlRn-bN5x:-KGrsdeXXPSeHLFUz-kJ:-216710473) ([raw file](https://github.com/servo/servo/blob/1138b89920ec9b8414551c0e311689258b4d601f/components/layout/wrapper.rs#L1360)):*
It's not really a programming error, since an unexpected UA selector could trigger this. Also, some methods are called (mainly `get_id` and `each_class`) when the selector map is queried for rules.

I went with `warn!()` since I didn't want UA stylesheets to make us crash, but if you prefer that (agreed it's more noticeable :P), I can do it.

What do you think about going with `warn!()` in the expected cases, and `error!()` in the others?

---

*[components/style/selector_impl.rs, line 117 \[r24\]](https://reviewable.io:443/reviews/servo/servo/10934#-KGrqGoZkEbEjmG6RvOz:-KGrtd1sfOXVG06vkVtD:-135522424) ([raw file](https://github.com/servo/servo/blob/1138b89920ec9b8414551c0e311689258b4d601f/components/style/selector_impl.rs#L117)):*
Done. I removed the note about `::selection` since it can't apply (it's a public pseudo and we wouldn't match as many selectors as we should).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10934)*
<!-- Sent from Reviewable.io -->

bholley commented:
FWIW, in Gecko there are plenty of ways to make layout crash by changing the UA stylesheet. We may or may not want to go that route in Servo, but it's not unheard-of.

mbrubeck commented:
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 1 of 1 files at r25.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[components/layout/wrapper.rs, line 1360 \[r24\]](https://reviewable.io:443/reviews/servo/servo/10934#-KGrotUZTmHMlRn-bN5x:-KGs2ECoSXShgbFrJXRg:830506266) ([raw file](https://github.com/servo/servo/blob/1138b89920ec9b8414551c0e311689258b4d601f/components/layout/wrapper.rs#L1360)):*
> What do you think about going with `warn!()` in the expected cases, and `error!()` in the others?

How about `debug!` for the expected cases and `warn!` for the others?  (I don't want any console spew in expected operations in release builds.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10934#-:-KGsFm-C8APGyhXJD-MQ:1458266033)*
<!-- Sent from Reviewable.io -->

emilio committed:
docs: Add style overview.

emilio committed:
layout: use `debug!` instead of `warn!` for the expected ThreadSafeLayoutElement methods

highfive commented:
New code was committed to pull request.

emilio commented:
Review status: 8 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[components/layout/wrapper.rs, line 1360 \[r24\]](https://reviewable.io:443/reviews/servo/servo/10934#-KGrotUZTmHMlRn-bN5x:-KGsM-xoftBpclYlZ7nV:-1667079951) ([raw file](https://github.com/servo/servo/blob/1138b89920ec9b8414551c0e311689258b4d601f/components/layout/wrapper.rs#L1360)):*
Fair enough, done :P

While waiting I wrote a style system overview, let me know if something is not clear enough or is incorrect :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10934)*
<!-- Sent from Reviewable.io -->

mbrubeck commented:
-S-awaiting-review +S-needs-code-changes 

Nice docs, thanks!  I think my review is complete.  r=mbrubeck with one typo fix (below).

---

Reviewed 4 of 4 files at r26.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/components/style.md, line 126 \[r26\]](https://reviewable.io:443/reviews/servo/servo/10934#-KGsXLL1PnacHfOf_riO:-KGsXLL1PnacHfOf_riP:1050534390) ([raw file](https://github.com/servo/servo/blob/c961f771c3b2653e9287dccbf73dc588c98eff6c/docs/components/style.md#L126)):*
typo: werther

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10934#-:-KGsXhB163gJbxPw4WVz:-653668412)*
<!-- Sent from Reviewable.io -->

emilio committed:
docs: Ammend the lazy pseudo-element section to be clearer

Thanks to Simon for the feedback :)

emilio committed:
geckolib: Implement Servo_GetComputedValuesForPseudoElement

highfive commented:
New code was committed to pull request.

emilio commented:
Ok, rebased, typo fixed, and implemented `Servo_GetComputedValuesForPseudoElement` for stylo.

r? @bholley (for the stylo changes, the rest is reviewed)

---

Review status: 9 of 13 files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/components/style.md, line 126 \[r26\]](https://reviewable.io:443/reviews/servo/servo/10934#-KGsXLL1PnacHfOf_riO:-KGsmm9Wd9z0Wbx2iMFR:-499146009) ([raw file](https://github.com/servo/servo/blob/c961f771c3b2653e9287dccbf73dc588c98eff6c/docs/components/style.md#L126)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10934#-:-KGsmnqxwIKd02zPdRU1:-1771340107)*
<!-- Sent from Reviewable.io -->

bholley commented on ports/geckolib/glue.rs:
```diff
@@ -267,12 +268,68 @@ pub extern "C" fn Servo_GetComputedValuesForAnonymousBox(parent_style_or_null: *
+        }
+    };
+
+    let pseudo = match pseudo_element_from_atom(pseudo_tag, /* ua_stylesheet = */ true) {
```
We'll want to pass the appropriate value here from the caller right? I'll add a TODO.

bholley commented:
Looks great!

@bors-servo r=bholley,mbrubeck

bors-servo commented:
:pushpin: Commit 0f7b70c has been approved by `bholley,mbrubeck`

<!-- @bors-servo r=bholley,mbrubeck 0f7b70c0b7836800ae44e714f7ac19ff50e1e34b -->

bors-servo commented:
:hourglass: Testing commit 0f7b70c with merge 04012c1...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/789)

highfive commented:
▶ TIMEOUT [expected OK] /html/rendering/replaced-elements/svg-embedded-sizing/svg-in-iframe-auto.html
      │ 
      └ Xlib:  extension &#34;XFree86-VidModeExtension&#34; missing on display &#34;:0&#34;.
    
      ▶ Unexpected subtest result in /html/rendering/replaced-elements/svg-embedded-sizing/svg-in-iframe-auto.html:
      │ TIMEOUT [expected PASS] placeholder: &#39;iframe&#39;, containerWidthStyle: &#39;400px&#39;, containerHeightStyle: &#39;400px&#39;, placeholderWidthAttr: &#39;100&#39;, svgViewBoxAttr: &#39;0 0 100 200&#39;, svgWidthAttr: &#39;25%&#39;, svgHeightAttr: &#39;25%&#39;, 
      └   → Test timed out
    
      ▶ TIMEOUT [expected OK] /html/rendering/replaced-elements/svg-embedded-sizing/svg-in-iframe-fixed.html
      │ 
      └ Xlib:  extension &#34;XFree86-VidModeExtension&#34; missing on display &#34;:0&#34;.
    
      ▶ Unexpected subtest result in /html/rendering/replaced-elements/svg-embedded-sizing/svg-in-iframe-fixed.html:
      │ TIMEOUT [expected PASS] placeholder: &#39;iframe&#39;, containerHeightStyle: &#39;400px&#39;, placeholderWidthAttr: &#39;100&#39;, placeholderHeightAttr: &#39;100px&#39;, svgViewBoxAttr: &#39;0 0 100 200&#39;, svgWidthAttr: &#39;25%&#39;, svgHeightAttr: &#39;25%&#39;, 
      └   → Test timed out
    
      ▶ TIMEOUT [expected OK] /html/rendering/replaced-elements/svg-embedded-sizing/svg-in-iframe-percentage.html
      │ 
      └ Xlib:  extension &#34;XFree86-VidModeExtension&#34; missing on display &#34;:0&#34;.
    
      ▶ Unexpected subtest result in /html/rendering/replaced-elements/svg-embedded-sizing/svg-in-iframe-percentage.html:
      │ TIMEOUT [expected PASS] placeholder: &#39;iframe&#39;, containerHeightStyle: &#39;400px&#39;, placeholderWidthAttr: &#39;100&#39;, placeholderHeightAttr: &#39;100%&#39;, svgViewBoxAttr: &#39;0 0 100 200&#39;, svgWidthAttr: &#39;25%&#39;, svgHeightAttr: &#39;25%&#39;, 
      └   → Test timed out

emilio commented:
@bors-servo: retry

https://github.com/servo/servo/issues/10457

bors-servo commented:
:hourglass: Testing commit 0f7b70c with merge 29823cb...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/703), [arm32](http://build.servo.org/builders/arm32/builds/751), [arm64](http://build.servo.org/builders/arm64/builds/736), [linux-dev](http://build.servo.org/builders/linux-dev/builds/750), [linux-rel](http://build.servo.org/builders/linux-rel/builds/802), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/764), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/744), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/781), [windows](http://build.servo.org/builders/windows/builds/436)
